### PR TITLE
Closes #188 | Implement dataloader for Sea-bench

### LINKED ISSUE #188: Create dataset loader for Sea-bench
Dataloader name: `seabench/seabench.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?seabench

|     Dataset| seabench |
|-------------|---|
| Description | Sea-bench is a multilingual benchmark for assistant-style models annotated by native linguists covering 8 Southeast Asian languages. The linguists sourced such data by manually translating open-source English test sets, collecting real user questions from local forums and websites, collecting real math and reasoning questions from reputable sources, as well as writing test instructions and questions themselves. The Sea-bench test set contains 20 questions per task (5 tasks for 3 languages, 4 tasks for other 5 languages). |
| Subsets | vi (Task-solving, Math-reasoning, General-instruction, NaturalQA, Safety), id (Task-solving, Math-reasoning, General-instruction, NaturalQA, Safety), th (Task-solving, Math-reasoning, General-instruction, NaturalQA, Safety), km (Task-solving, Math-reasoning, General-instruction, NaturalQA), lo (Task-solving, Math-reasoning, General-instruction, NaturalQA), ms (Task-solving, Math-reasoning, General-instruction, NaturalQA), my (Task-solving, Math-reasoning, General-instruction, NaturalQA), tl (Task-solving, Math-reasoning, General-instruction, NaturalQA) |
| Languages | tgl, khm, vie, tha, lao, mya, ind, zlm |
| Tasks | Instruction Tuning |
| License | Apache license 2.0 (apache-2.0) |
| Homepage | https://huggingface.co/datasets/SeaLLMs/Sea-bench/raw/main/question.jsonl |
| HF URL | https://huggingface.co/datasets/SeaLLMs/Sea-bench |
| Paper URL | https://arxiv.org/pdf/2312.00738.pdf |

### PR DESCRIPTION
**Pull request by raileymontalan**
Closes [#188 ](https://github.com/SEACrowd/seacrowd-datahub/issues/188).

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/wit/wit.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/wit/wit.py`.
- [x] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

**Comment by MJonibek** [comment]
Hello @raileymontalan, thank you for working on this dataloader. Can you please leave only the seabench dataloader file in this PR. Right now you have some files that are not related to this PR. Also please add `__init__.py` file for the sea_bench dataloader. 

Please run `make check_file=seacrowd/sea_datasets/sea_bench/sea_bench.py` command to format python file. 

Regarding sea_bench dataloader implementation, it works fine, but I am not sure if is it correctly implemented. For example, for QA schemas there are samples without answers.

**Review by MJonibek — CHANGES_REQUESTED** [review]
Hello @raileymontalan, thank you for working on this dataloader. Can you please leave only the seabench dataloader file in this PR. Right now you have some files that are not related to this PR. Also please add __init__.py file for the sea_bench dataloader.

Please run make check_file=seacrowd/sea_datasets/sea_bench/sea_bench.py command to format python file.

Regarding sea_bench dataloader implementation, it works fine, but I am not sure if is it correctly implemented. For example, for QA schemas there are samples without answers.

**Review by jen-santoso — CHANGES_REQUESTED** [review]
Hello @raileymontalan 

Thank you for implementing the dataloader! Tested and runs fine. 
There are a few revisions needed here (some overlaps with @MJonibek 's review), please refer to the comment below and in the code:

- Include the `__init__.py` file for this dataloader
- Remove files unrelated to this PR. Only keep the ones for `sea_bench`.
- Please fix the small nits and run make check_file before pushing the PR update.

Waiting for your updates!

**Comment by sabilmakbar** [comment]
Quick check: Mind removing the `wit` dataloader from this PR, @raileymontalan? Thx!

**Comment by holylovenia** [comment]
Hi, recently I came across Sea-bench's datasheet in SEACrowd and wonder if instruction tuning would suit this dataloader better compared to QA and summarization. I would like to suggest to use these tasks for the dataloader: `Tasks.INSTRUCTION_TUNING` and `Tasks.MACHINE_TRANSLATION`. What do you think, @raileymontalan @jen-santoso @MJonibek?

cc: @SamuelCahyawijaya, is there any specific reason why Sea-bench was originally submitted with summarization, QA, and MT tasks?

**Comment by holylovenia** [comment]
Hi, recently I came across Sea-bench's datasheet in SEACrowd and wonder if instruction tuning would suit this dataloader better compared to QA and summarization. I would like to suggest to use these tasks for the dataloader: `Tasks.INSTRUCTION_TUNING` and `Tasks.MACHINE_TRANSLATION`. What do you think, @raileymontalan @jen-santoso @MJonibek?

cc: @rmahendra, is there any specific reason why Sea-bench was originally submitted with summarization, QA, and MT tasks?

Also tagging @SamuelCahyawijaya just in case he has any opinion on this.

File: seacrowd/sea_datasets/sea_bench/sea_bench.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+collecting real math and reasoning questions from reputable sources, as well as writing test
+instructions and questions themselves. The Sea-bench test set contains 20 questions per task
+(5 tasks for 3 languages, 4 tasks for other 5 languages).
+"""
+
+_HOMEPAGE = "https://huggingface.co/datasets/SeaLLMs/Sea-bench"
+
+_LANGUAGES = {"tgl": "tl", "khm": "km", "vie": "vi", "tha": "th", "lao": "lo", "mya": "my", "ind": "id", "zlm": "ms", "eng": "en"}
+
+_LANGUAGE_CODES = list(_LANGUAGES.values())
```
[the remark this answers is not in the document]
Comment (raileymontalan) [reply seacrowd/sea_datasets/sea_bench/sea_bench.py]: Removed.

File: seacrowd/sea_datasets/sea_bench/sea_bench.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+_URLS = "https://huggingface.co/datasets/SeaLLMs/Sea-bench/raw/main/question.jsonl"
+
+_SUPPORTED_TASKS = [Tasks.SUMMARIZATION, Tasks.MACHINE_TRANSLATION, Tasks.QUESTION_ANSWERING]
+
+_SOURCE_VERSION = "1.0.0"
+
+_SEACROWD_VERSION = "1.0.0"
+
+logger = datasets.logging.get_logger(__name__)
```
[the remark this answers is not in the document]
Comment (raileymontalan) [reply seacrowd/sea_datasets/sea_bench/sea_bench.py]: Removed.

**Comment by MJonibek** [comment]
@holylovenia Yes, I agree that `Tasks.INSTRUCTION_TUNING` suits better, but I am not sure about `Tasks.MACHINE_TRANSLATION`. Is the dataset parallel, because there is no machine translation in the category list? If yes, how to make sure that instructions in different languages are the same?

**Comment by holylovenia** [comment]
> @holylovenia Yes, I agree that `Tasks.INSTRUCTION_TUNING` suits better, but I am not sure about `Tasks.MACHINE_TRANSLATION`. Is the dataset parallel, because there is no machine translation in the category list? If yes, how to make sure that instructions in different languages are the same?

On a second thought, I agree with you. Let's just implement `Tasks.INSTRUCTION_TUNING` since we only implemented the tasks intended by the dataset.

@raileymontalan @jen-santoso, does this sound okay to you? If yes, I'll also modify the datasheet.

**Comment by jen-santoso** [comment]
> > @holylovenia Yes, I agree that `Tasks.INSTRUCTION_TUNING` suits better, but I am not sure about `Tasks.MACHINE_TRANSLATION`. Is the dataset parallel, because there is no machine translation in the category list? If yes, how to make sure that instructions in different languages are the same?
> 
> On a second thought, I agree with you. Let's just implement `Tasks.INSTRUCTION_TUNING` since we only implemented the tasks intended by the dataset.
> 
> @raileymontalan @jen-santoso, does this sound okay to you? If yes, I'll also modify the datasheet.

@holylovenia This sound OK to me.

**Comment by raileymontalan** [comment]
@jen-santoso @MJonibek I've applied the requested changes and removed the machine translation task. Thanks!

**Review by MJonibek — COMMENTED** [review]
@raileymontalan Checked, LGTM. One concern: there are None in the text_2 column (380 out of 780 samples without chatgpt response). Also, all text_1 column values are in the form of: "["....value..."]". Can you remove these square brackets?

File: seacrowd/sea_datasets/sea_bench/sea_bench.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+_URLS = "https://huggingface.co/datasets/SeaLLMs/Sea-bench/raw/main/question.jsonl"
+
+_SUPPORTED_TASKS = [Tasks.INSTRUCTION_TUNING]
+
+_SOURCE_VERSION = "1.0.0"
+
+_SEACROWD_VERSION = "1.0.0"
+
+
+class WITDataset(datasets.GeneratorBasedBuilder):
```
Comment (jen-santoso) [inline seacrowd/sea_datasets/sea_bench/sea_bench.py:66]:
```suggestion
class SeaBench(datasets.GeneratorBasedBuilder):
```

File: seacrowd/sea_datasets/sea_bench/sea_bench.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+covering 8 Southeast Asian languages. The linguists sourced such data by manually translating
+open-source English test sets, collecting real user questions from local forums and websites,
+collecting real math and reasoning questions from reputable sources, as well as writing test
+instructions and questions themselves. The Sea-bench test set contains 20 questions per task
+(5 tasks for 3 languages, 4 tasks for other 5 languages).
+"""
+
+_HOMEPAGE = "https://huggingface.co/datasets/SeaLLMs/Sea-bench"
+
+_LANGUAGES = {"tgl": "tl", "khm": "km", "vie": "vi", "tha": "th", "lao": "lo", "mya": "my", "ind": "id", "zlm": "ms", "eng": "en"}
```
Comment (jen-santoso) [inline seacrowd/sea_datasets/sea_bench/sea_bench.py:51]: Keep the `_LANGUAGES` containing list of the 3-code languages. Move the dict into new constant / variable.

**Review by jen-santoso — CHANGES_REQUESTED** [review]
Tested and runs well. Few nits to fix and all good to go

File: seacrowd/sea_datasets/sea_bench/sea_bench.py (hunk 3 above)
Comment (raileymontalan) [reply seacrowd/sea_datasets/sea_bench/sea_bench.py:66]: Will fix in the next commit. Thanks!

File: seacrowd/sea_datasets/sea_bench/sea_bench.py (hunk 4 above)
Comment (raileymontalan) [reply seacrowd/sea_datasets/sea_bench/sea_bench.py:51]: Will fix in the next commit. Thanks!

**Comment by raileymontalan** [comment]
> @raileymontalan Checked, LGTM. One concern: there are None in the text_2 column (380 out of 780 samples without chatgpt response). Also, all text_1 column values are in the form of: "["....value..."]". Can you remove these square brackets?

Hi @MJonibek, in the [original dataset](https://huggingface.co/datasets/SeaLLMs/Sea-bench), 380 (~48.7%) of the `chatgpt_response` rows really have null values, hence the `None`s. Would you prefer if those `None`s were just empty strings `""` instead? 

Also noted on the `text_1` column values. I've updated the code accordingly.

**Comment by MJonibek** [comment]
@raileymontalan Yeah, I see. Let`s leave it as it is. Maybe instructions without a response will be useful for users also. What do you think @jen-santoso?

**Review by MJonibek — APPROVED** [review]

**Comment by jen-santoso** [comment]
@MJonibek  I guess it is OK to leave the no-response as is.
@raileymontalan Thanks for the fix! LGTM!

**Review by jen-santoso — APPROVED** [review]

**Pull request merged**